File: Cs/Solution/SurfaceArea.cs
using System;

namespace Solution
{
	public partial class Solution
	{
		public int SurfaceArea(int[][] grid)
		{
			int area = 0;
			int n = grid.Length;
			int[][] dirs = { new int[] { 0, 1 }, new int[] { 0, -1 }, new int[] { 1, 0 }, new int[] { -1, 0 } };
			for (int i = 0; i < n; i++)
			{
				for (int j = 0; j < n; j++)
				{
					if (grid[i][j] > 0)
					{
						area += 2; /* each tower has top and bottom */
						for (int k = 0; k < 4; k++)
						{
							int x = i + dirs[k][0];
							int y = j + dirs[k][1];
							int v = 0;
							if (x >= 0 && x < n && y >= 0 && y < n)
							{
								v = grid[x][y];
							}
							/* if the neighbor is higher, then the area of the neighbor is hidden */
							area += Math.Max(0, grid[i][j] - v);

						}
					}
				}
			}
			return area;
		}
	}
}

File: Ruby/solution/surface_area.rb
# @param {Integer[][]} grid
# @return {Integer}
def surface_area(grid)
  area = 0
  rows = grid.size
  cols = grid[0].size

  (0...rows).each do |i|
    (0...cols).each do |j|
      height = grid[i][j]
      next if height == 0

      # Top and bottom faces
      area += 2

      # Front face
      area += height if i == 0
      area += [height - grid[i - 1][j], 0].max if i > 0

      # Back face
      area += height if i == rows - 1
      area += [height - grid[i + 1][j], 0].max if i < rows - 1

      # Left face
      area += height if j == 0
      area += [height - grid[i][j - 1], 0].max if j > 0

      # Right face
      area += height if j == cols - 1
      area += [height - grid[i][j + 1], 0].max if j < cols - 1
    end
  end

  area
end
